### DISCUSSION TITLE
Problem when starting osgEarth

### DISCUSSION THREAD

**Discussion by XianranZ** [pelicanmapping/osgearth#2650] [Support Q&A] [answered]
hello, everyone!
I compiled osgEarth3.7.1 with the source code. And then I wrote a demo to check if it works.
Here is my code:
`#include <osgViewer/Viewer>
#include <osgDB/ReadFile>
#include <osgUtil/Optimizer>
#include <osgEarth/MapNode>

int main()
{
    osg::ref_ptr<osgViewer::Viewer> viewer = new osgViewer::Viewer;

    osg::ref_ptr<osg::Group> root = new osg::Group;

    osg::ref_ptr<osg::Node> node = osgDB::readNodeFile("/home/zxr/Code/tesosgearth/simple.earth");

    osg::ref_ptr<osgEarth::MapNode> mapNode = osgEarth::MapNode::findMapNode(node.get());

    root->addChild(mapNode.get());

    osgUtil::Optimizer optimize;
    optimize.optimize(root.get());

    viewer->setSceneData(root.get());

    viewer->realize();

    viewer->run();

    return 0;
}`
The code compiles successfully but fails to run. 
Here is the error message:
[osgEarth]* ASSERTION FAILURE (createMains @ ShaderFactory.cpp:170) state.getUseVertexAttributeAliasing() ...OpenSceneGraph vertex attribute aliasing must be enabled. Consider installing the GL3RealizeOperation in your viewer.
[osgEarth]* ASSERTION FAILURE (createMains @ ShaderFactory.cpp:173) state.getUseModelViewAndProjectionUniforms() ...OpenSceneGraph matrix uniforms must be enabled. Consider installing the GL3RealizeOperation in your viewer.
VERTEX glCompileShader "oe_rex_init_model" FAILED
VERTEX Shader "oe_rex_init_model" infolog:
0:24(19): error: `gl_MultiTexCoord0' undeclared
0:30(32): error: `gl_MultiTexCoord0' undeclared
0:30(32): error: type mismatch
0:30(28): error: cannot construct `int' from a non-numeric data type

VERTEX glCompileShader "oe_rex_normalMapVS" FAILED
VERTEX Shader "oe_rex_normalMapVS" infolog:
0:30(17): warning: `oe_layer_tilec' used uninitialized
0:35(33): error: `gl_NormalMatrix' undeclared
0:35(33): error: operands to arithmetic operators must be numeric
0:35(23): error: no matching function for call to `normalize(error)'; candidates are:
0:35(23): error:    float normalize(float)
0:35(23): error:    vec2 normalize(vec2)
0:35(23): error:    vec3 normalize(vec3)
0:35(23): error:    vec4 normalize(vec4)
0:35(23): error:    double normalize(double)
0:35(23): error:    dvec2 normalize(dvec2)
0:35(23): error:    dvec3 normalize(dvec3)
0:35(23): error:    dvec4 normalize(dvec4)

VERTEX glCompileShader "oe_rex_morph" FAILED
VERTEX Shader "oe_rex_morph" infolog:
0:48(40): warning: `oe_layer_tilec' used uninitialized
0:51(29): error: `gl_ModelViewMatrix' undeclared
0:51(29): error: operands to arithmetic operators must be numeric
0:57(32): warning: `wouldBePositionView' used uninitialized
0:66(7): warning: `oe_terrain_vertexMarker' used uninitialized
0:68(62): warning: `vp_Normal' used uninitialized
0:70(34): error: `gl_MultiTexCoord1' undeclared
0:70(34): error: type mismatch
0:70(29): error: cannot construct `vec4' from a non-numeric data type
0:71(24): error: `gl_MultiTexCoord2' undeclared
0:71(24): error: type mismatch
0:76(30): warning: `oe_layer_tilec' used uninitialized
0:77(24): warning: `oe_layer_tilec' used uninitialized
0:78(28): warning: `oe_layer_tilec' used uninitialized
0:80(41): warning: `neighborVertexModel' used uninitialized
0:82(28): warning: `vp_Normal' used uninitialized
0:82(39): warning: `neighborNormal' used uninitialized
0:83(26): warning: `oe_layer_tilec' used uninitialized

VERTEX glCompileShader "main(vertex)" FAILED
VERTEX Shader "main(vertex)" infolog:
0:48(14): error: `gl_Vertex' undeclared
0:49(14): error: `gl_Normal' undeclared
0:50(13): error: `gl_Color' undeclared
0:54(14): error: `gl_ModelViewMatrix' undeclared
0:54(14): error: operands to arithmetic operators must be numeric
0:55(24): error: `gl_NormalMatrix' undeclared
0:55(24): error: operands to arithmetic operators must be numeric
0:55(14): error: no matching function for call to `normalize(error)'; candidates are:
0:55(14): error:    float normalize(float)
0:55(14): error:    vec2 normalize(vec2)
0:55(14): error:    vec3 normalize(vec3)
0:55(14): error:    vec4 normalize(vec4)
0:55(14): error:    double normalize(double)
0:55(14): error:    dvec2 normalize(dvec2)
0:55(14): error:    dvec3 normalize(dvec3)
0:55(14): error:    dvec4 normalize(dvec4)
0:62(16): error: `gl_ProjectionMatrix' undeclared
0:62(16): error: operands to arithmetic operators must be numeric
0:63(35): warning: `oe_terrain_vertexMarker' used uninitialized
0:64(28): warning: `oe_layer_opacity' used uninitialized
0:65(30): warning: `oe_rex_morphFactor' used uninitialized
0:66(25): warning: `oe_layer_texc' used uninitialized
0:67(31): warning: `oe_layer_texcParent' used uninitialized
0:68(24): warning: `oe_normal_uv' used uninitialized
0:69(27): warning: `oe_UpVectorView' used uninitialized
0:70(30): warning: `oe_normal_binormal' used uninitialized
0:73(26): warning: `oe_layer_tilec' used uninitialized

[osgEarth]* [VirtualProgram] Program will not link!
[osgEarth]* ASSERTION FAILURE (createMains @ ShaderFactory.cpp:170) state.getUseVertexAttributeAliasing() ...OpenSceneGraph vertex attribute aliasing must be enabled. Consider installing the GL3RealizeOperation in your viewer.
[osgEarth]* ASSERTION FAILURE (createMains @ ShaderFactory.cpp:173) state.getUseModelViewAndProjectionUniforms() ...OpenSceneGraph matrix uniforms must be enabled. Consider installing the GL3RealizeOperation in your viewer.
VERTEX glCompileShader "main(vertex)" FAILED
VERTEX Shader "main(vertex)" infolog:
0:47(14): error: `gl_Vertex' undeclared
0:48(14): error: `gl_Normal' undeclared
0:49(13): error: `gl_Color' undeclared
0:52(14): error: `gl_ModelViewMatrix' undeclared
0:52(14): error: operands to arithmetic operators must be numeric
0:53(24): error: `gl_NormalMatrix' undeclared
0:53(24): error: operands to arithmetic operators must be numeric
0:53(14): error: no matching function for call to `normalize(error)'; candidates are:
0:53(14): error:    float normalize(float)
0:53(14): error:    vec2 normalize(vec2)
0:53(14): error:    vec3 normalize(vec3)
0:53(14): error:    vec4 normalize(vec4)
0:53(14): error:    double normalize(double)
0:53(14): error:    dvec2 normalize(dvec2)
0:53(14): error:    dvec3 normalize(dvec3)
0:53(14): error:    dvec4 normalize(dvec4)
0:60(16): error: `gl_ProjectionMatrix' undeclared
0:60(16): error: operands to arithmetic operators must be numeric
0:61(35): warning: `oe_terrain_vertexMarker' used uninitialized
0:62(28): warning: `oe_layer_opacity' used uninitialized
0:63(30): warning: `oe_rex_morphFactor' used uninitialized
0:64(25): warning: `oe_layer_texc' used uninitialized
0:65(31): warning: `oe_layer_texcParent' used uninitialized
0:66(24): warning: `oe_normal_uv' used uninitialized
0:67(27): warning: `oe_UpVectorView' used uninitialized
0:68(30): warning: `oe_normal_binormal' used uninitialized
0:71(26): warning: `oe_layer_tilec' used uninitialized

[osgEarth]* [VirtualProgram] Program will not link!
glLinkProgram 0x7f87d8029ba0"GDALImageLayer" FAILED
Program "GDALImageLayer" infolog:
error: linking with uncompiled/unspecialized shadererror: linking with uncompiled/unspecialized shadererror: linking with uncompiled/unspecialized shadererror: linking with uncompiled/unspecialized shader
glLinkProgram 0x7f87d862c5d0"Terrain image layer" FAILED
Program "Terrain image layer" infolog:
error: linking with uncompiled/unspecialized shadererror: linking with uncompiled/unspecialized shadererror: linking with uncompiled/unspecialized shadererror: linking with uncompiled/unspecialized shader
Does anyone knows why?

runtime environment:
OS: ubuntu22
IDE: Qt6

the content of cmakelists.txt
cmake_minimum_required(VERSION 3.16)

project(tesosgearth LANGUAGES CXX)

set(CMAKE_CXX_STANDARD 17)
set(CMAKE_CXX_STANDARD_REQUIRED ON)

find_package(osg REQUIRED)
find_package(osgViewer REQUIRED)
find_package(osgDB REQUIRED)
find_package(osgUtil REQUIRED)
find_package(OpenGL REQUIRED)

set(OSGEARTH_LIBRARY "/usr/local/lib64/libosgEarth.so" "/usr/local/lib64/libosgEarthImGui.so")

add_executable(tesosgearth main.cpp)

target_link_libraries(tesosgearth osg osgViewer osgDB osgUtil OpenGL ${OSGEARTH_LIBRARY})

include(GNUInstallDirs)
install(TARGETS tesosgearth
    LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
    RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
)

if you can help me, thank a lot!

**Comment by gwaldron** [comment, marked as answer]
The assertion error offers a suggestion:
```
[osgEarth]* ASSERTION FAILURE ... Consider installing the GL3RealizeOperation in your viewer.
```
Try that.
```c++
#include <osgEarth/GLUtils>
...
viewer.setRealizeOperation(new GL3RealizeOperation());
```

**Reply by XianranZ** [reply]
Thanks a lot! It's useful.

**Reply by XianranZ** [reply]
And I have a new question. In the old version of osgEarth, there is a tool named osgearth_package to tile the big image. But in the newest osgEarth I compiled, the osgearth_package is disappeared. I want to know if the developer removed the tool or if it changed its name, or if there is something wrong with my compilation.

**Reply by gwaldron** [reply]
`osgearth_package` was removed long ago.
You can use `osgearth_conv` to create local tiled datasets. There are some docs here:
https://docs.osgearth.org/en/latest/data.html#building-tilesets

**Reply by XianranZ** [reply]
thanks a lot! it is very helpful
